src/services/release.ts

```diff
@@ -104,127 +104,137 @@ export async function calculateRelease(): Promise<void> {
-      logger.error(
-        `Some of the expected files to publish were discovered as no longer PUBLISHED in the DataCenter: ${missingFileIds}`,
-      );
-      throw new Error(
-        `Some files expected in the release were not retrieved when fetching data to buld the Public Index. \
+	try {
+		logger.info(`Beginning release building...`);
+		let release = await releaseService.getActiveRelease();
+		if (!release) {
+			throw new Error('No Active release available.');
+		}
+
+		try {
+			release = await releaseService.updateActiveReleaseLabel(label);
+		} catch (e) {
+			logger.debug(JSON.stringify(e));
+			throw new Error(`Release with the label '${label}' already exists`);
+		}
+
+		// 1. Sort files into programs, published and restricted
+		const programs: Record<string, { kept: File[]; added: File[] }> = {};
+		const filesKept: File[] = await fileService.getFilesByObjectIds(release.filesKept);
+		const filesAdded: File[] = await fileService.getFilesByObjectIds(release.filesAdded);
+		const filesRemoved: File[] = await fileService.getFilesByObjectIds(release.filesRemoved);
+
+		// programIds is used to list all public indices that should be created.
+		// including filesRemoved in this list ensures that if we remove all files from a program's public index
+		// then an empty index is created to replace the existing one.
+		const programIds = new Set<string>();
+		[...filesKept, ...filesAdded, ...filesRemoved].forEach(file => programIds.add(file.programId));
+
+		filesKept.forEach(file => {
+			const program = file.programId;
+			if (!programs[program]) {
+				programs[program] = { kept: [], added: [] };
+			}
+			programs[program].kept.push(file);
+		});
+		filesAdded.forEach(file => {
+			const program = file.programId;
+			if (!programs[program]) {
+				programs[program] = { kept: [], added: [] };
+			}
+			programs[program].added.push(file);
+		});
+
+		// 2. Create public indices for each program (no clone!)
+		const indexer = await getIndexer();
+
+		// 2.a If the release already has public indices, remove those so we can build from the existing indices.
+		await indexer.deleteIndices(release.indices);
+		release = await releaseService.updateActiveReleaseIndices([]);
+
+		// 2.b Make new empty public indices from the current public indices.
+		const publicIndices = await indexer.createEmptyPublicIndices(Array.from(programIds));
+		release = await releaseService.updateActiveReleaseIndices(publicIndices);
+
+		// 3. Add files all required files to public indices.
+		const expectedPublicFiles = filesAdded.concat(filesKept);
+
+		// 3a. Update public files with latest clinical data status
+		const { results: updatedExpectedPublicFiles } = await PromisePool.withConcurrency(5)
+			.for(expectedPublicFiles)
+			.handleError((error, file) => {
+				logger.error(`Failed to update clinical data status in DB for ${file}: ${error}`);
+			})
+			.process(async file => {
+				return await fileManager.updateFileClinicalData(file);
+			});
```

Comment:
This is the only code change made to `services/release.ts`. This is the work required to update the clinical data for every file that will be made public. It is being done with some minimal concurrency to try to speed up the process. All the actual work is performed by the existing function `fileManager.updateFileClinicalData`.